soci/ztoc_builder.go

```diff
@@ -81,34 +83,113 @@ func BuildZtoc(gzipFile string, span int64, cfg *buildConfig) (*Ztoc, error) {
 }
 
 func NewZtocReader(ztoc *Ztoc) (io.Reader, ocispec.Descriptor, error) {
-	serializedBuf := new(bytes.Buffer)
-	enc := gob.NewEncoder(serializedBuf)
-	err := enc.Encode(*ztoc)
-	if err != nil {
-		return nil, ocispec.Descriptor{}, fmt.Errorf("cannot serialize ztoc: %w", err)
-	}
+	flatbuf := ztocToFlatbuffer(ztoc)
+	buf := bytes.NewReader(flatbuf)
+	dgst := digest.FromBytes(flatbuf)
+	size := len(flatbuf)
+	return buf, ocispec.Descriptor{
+		Digest: dgst,
+		Size:   int64(size),
+	}, nil
+}
 
-	compressedBuf := new(bytes.Buffer)
-	zs, err := zstd.NewWriter(compressedBuf)
-	if err != nil {
-		return nil, ocispec.Descriptor{}, fmt.Errorf("cannot create zstd writer: %w", err)
+func ztocToFlatbuffer(ztoc *Ztoc) []byte {
+	builder := flatbuffers.NewBuilder(0)
+	version := builder.CreateString(ztoc.Version)
+	buildToolIdentifier := builder.CreateString(ztoc.BuildToolIdentifier)
+
+	metadataOffsetList := make([]flatbuffers.UOffsetT, len(ztoc.Metadata))
+	for i := len(ztoc.Metadata) - 1; i >= 0; i-- {
+		me := ztoc.Metadata[i]
```

Comment:
Better to extract a few functions from the loop.

Reply:
Done.